aspects/schema_test.go

```diff
@@ -1680,3 +1682,154 @@ func (*schemaSuite) TestArrayFailsWithBadUniqueType(c *C) {
+			input: []byte(`{"bar": 1}`),
+			err:   `cannot validate top level element: map contains unexpected key "bar"`,
+		},
+		{
+			name:  "1 level of nesting",
+			input: []byte(`{"foo": {"baz": 1}}`),
+			err:   `cannot validate element under "foo": map contains unexpected key "baz"`,
+		},
+		{
+			name:  "2 levels of nesting",
+			input: []byte(`{"foo": {"bar": {"boo": 1}}}`),
+			err:   `cannot validate element under "foo.bar": map contains unexpected key "boo"`,
+		},
+	}
+
+	for _, tc := range testcases {
+		err = schema.Validate(tc.input)
+		c.Assert(err, ErrorMatches, tc.err, Commentf("test case %q failed", tc.name))
+	}
+}
+
+func (*schemaSuite) TestPathPrefixWithMapUnderUserType(c *C) {
+	schemaStr := []byte(`{
+	"types": {
+		"my-type": {
+			"schema": {
+				"bar": {
+					"type": "int",
+					"min": 0
+				}
+			}
+		}
+	},
+	"schema": {
+		"foo": "$my-type"
+	}
+}`)
+
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": {"bar": -1}}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo.bar": -1 is less than the allowed minimum 0`)
+}
+
+func (*schemaSuite) TestPathPrefixWithArrayUnderUserType(c *C) {
+	schemaStr := []byte(`{
+	"types": {
+		"my-type": {
+			"type": "int",
+			"min": 0
+		}
+	},
+	"schema": {
+		"foo": {
+			"type": "array",
+			"values": "$my-type"
+		}
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": [-1]}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo": -1 is less than the allowed minimum 0`)
+}
+
+func (*schemaSuite) TestPathPrefixWithArrayUnderUserWithAContainerElementType(c *C) {
+	schemaStr := []byte(`{
+	"types": {
+		"my-type": {
+			"type": "array",
+			"values": {
+				"schema": {
+					"bar": {
+						"type": "int",
+						"min": 0
+					}
+				}
+			}
+		}
+	},
+	"schema": {
+		"foo": "$my-type"
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": [{"bar": -1}]}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo.bar": -1 is less than the allowed minimum 0`)
+}
+
+func (*schemaSuite) TestPathPrefixWithKeyOrValueConstraints(c *C) {
+	schemaStr := []byte(`{
+	"schema": {
+		"foo": {
+			"type": "map",
+			"keys": {
+				"type": "string",
+				"choices": ["my-key"]
+			},
+			"values": {
+				"type": "int",
+				"min": 0
+			}
+		}
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": {"other-key": 1}}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo": string "other-key" is not one of the allowed choices`)
+
+	input = []byte(`{"foo": {"my-key": -1}}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo": -1 is less than the allowed minimum 0`)
```

Comment:
General nitpick. "cannot validate" means the inability to verify, which is not really the case.